content/2016-03-29-compiler-flags.rst

```diff
@@ -0,0 +1,275 @@
+one. This slightly impacts the code size and eventually gives the compiler more
+optimization opportunities, but nothing significant there (131K), apart the pleasure
+of generating cleaner binaries. That's also going to be the default for next
+Pythran version.
+
+Out of chance: getting faster binaries
+======================================
+
+In the (huge) info pages of GCC, near the doc of ``-fvisibility=hidden``,
+there's this (GCC only) compiler flag, ``-fwhole-program`` that implements some
+kind of Link Time Optimization, in the sense that it tells the compiler to
+consider the current compilation unit (or code) as a whole program. ``main`` is
+still the entry point, but one can set extra attributes to function to flag
+them as exported too, using ``__attribute__ ((externally_visible))``. That
+sounds a bit redundant to me with the visibility attribute, but it turns out
+this triggers abunch of different optimization path that gives us a significantly smaller
+binary, that runs slightly faster:
+
+.. code:: sh
+
+    > pythran cda.py -fvisibility=hidden -fwhole-program -Wl,-strip-all
+    > ls -lh cda.so
+    -rwxr-xr-x 1 sguelton sguelton 31K Mar 29 18:52 cda.so*
+    > python -m timeit -s 'import numpy as np; n = 20000 ; lat, lon = np.random.rand(n), np.random.rand(n); x,y = np.random.rand(), np.random.rand(); from cda import closest_distance_arrays' 'closest_distance_arrays(x,y,lat, lon)'
+    1000 loops, best of 3: 1.15 msec per loop
+
+All these flags are now the default on Linux.
+
+Playing with the optimization flags too
+=======================================
+
+The default optimization flag is ``-O2``, and that's generally a decent choice.
+On ``cda.py``, using ``-O3`` does not give much change (gcc 4.9):
+
+.. code:: sh
+
+    > pythran cda.py -fvisibility=hidden -fwhole-program -Wl,-strip-all -O3
+    > python -m timeit [...]
+    1000 loops, best of 3: 1.14 msec per loop
+
+Asking for code specific to my CPU using ``-march=native`` actually gives some improvments
+
+.. code:: sh
+
+    > pythran cda.py -fvisibility=hidden -fwhole-program -Wl,-strip-all -O3 -march=native
+    > python -m timeit [...]
+    1000 loops, best of 3: 1.11 msec per loop
+
+But the best speedup has a price: relaxing standard compliance with ``-Ofast``
+can be beneficial if you're not using denormalized numbers, infinity and the
+monstrosity that lies with ``NaN``:
+
+.. code:: sh
+
+    > pythran cda.py -fvisibility=hidden -fwhole-program -Wl,-strip-all -Ofast -march=native
+    > python -m timeit [...]
+    1000 loops, best of 3: 1.02 msec per loop
+
+If you're really into compiler flags tuning, you can try out ``-funroll-loops``
+or try to tune the ``-finline-limit=N`` parameter (that actually get mets dow
+to ``1ms per loop``) but that's going a bit too far :-)
+
+Don't forget Vectorization
+==========================
+
+Combining ``-O3`` and ``-march=native`` triggers compiler auto-vectorization,
```

Comment:
only GCC needs this, clang turns vectorisation at -O2 (SSE2 IIRC w/ x86). -march=native will allow him to use a more recent instruction set if available